File: internal/db/models/vm/vm.go
// SPDX-License-Identifier: BSD-2-Clause
//
// Copyright (c) 2025 The FreeBSD Foundation.
//
// This software was developed by Hayzam Sherif <hayzam@alchemilla.io>
// of Alchemilla Ventures Pvt. Ltd. <hello@alchemilla.io>,
// under sponsorship from the FreeBSD Foundation.

package vmModels

import (
	"time"

	networkModels "github.com/alchemillahq/sylve/internal/db/models/network"
)

type Storage struct {
	ID        uint   `gorm:"primaryKey" json:"id"`
	Name      string `json:"name" gorm:"default:''"`
	Type      string `json:"type"`
	Dataset   string `json:"dataset"`
	Size      int64  `json:"size"`
	Emulation string `json:"emulation"`

	VMID uint `json:"vmId" gorm:"index"`
}

type Network struct {
	ID  uint   `gorm:"primaryKey" json:"id"`
	MAC string `json:"mac"`

	MacID      *uint                 `json:"macId" gorm:"column:mac_id"`
	AddressObj *networkModels.Object `json:"macObj" gorm:"foreignKey:MacID"`

	SwitchID uint                         `json:"switchId" gorm:"not null;index"`
	Switch   networkModels.StandardSwitch `gorm:"foreignKey:SwitchID;constraint:OnUpdate:CASCADE,OnDelete:RESTRICT"`

	Emulation string `json:"emulation"`

	VMID uint `json:"vmId" gorm:"index"`
}

type VMStats struct {
	ID          uint    `gorm:"primaryKey" json:"id"`
	VMID        uint    `json:"vmId" gorm:"index"`
	CPUUsage    float64 `json:"cpuUsage"`
	MemoryUsage float64 `json:"memoryUsage"`
	MemoryUsed  float64 `json:"memoryUsed"`

	CreatedAt time.Time `json:"createdAt" gorm:"autoCreateTime"`
}

type VM struct {
	ID            uint   `gorm:"primaryKey" json:"id"`
	Name          string `json:"name"`
	Description   string `json:"description"`
	VmID          int    `json:"vmId"`
	CPUSockets    int    `json:"cpuSockets"`
	CPUCores      int    `json:"cpuCores"`
	CPUsThreads   int    `json:"cpuThreads"`
	RAM           int    `json:"ram"`
	VNCPort       int    `json:"vncPort"`
	VNCPassword   string `json:"vncPassword"`
	VNCResolution string `json:"vncResolution"`
	VNCWait       bool   `json:"vncWait"`
	StartAtBoot   bool   `json:"startAtBoot"`
	TPMEmulation  bool   `json:"tpmEmulation"`
	StartOrder    int    `json:"startOrder"`
	WoL           bool   `json:"wol" gorm:"default:false"`

	ISO        string    `json:"iso"`
	Storages   []Storage `json:"storages" gorm:"foreignKey:VMID;references:ID;constraint:OnUpdate:CASCADE,OnDelete:CASCADE;"`
	Networks   []Network `json:"networks" gorm:"foreignKey:VMID;references:ID;constraint:OnUpdate:CASCADE,OnDelete:CASCADE;"`
	PCIDevices []int     `json:"pciDevices" gorm:"serializer:json;type:json"`
	CPUPinning []int     `json:"cpuPinning" gorm:"serializer:json;type:json"`

	Stats []VMStats `json:"-" gorm:"foreignKey:VMID;references:ID;constraint:OnUpdate:CASCADE,OnDelete:CASCADE;"`
	State string    `json:"state" gorm:"-"`

	CreatedAt time.Time `json:"createdAt" gorm:"autoCreateTime"`
	UpdatedAt time.Time `json:"updatedAt" gorm:"autoUpdateTime"`

	StartedAt *time.Time `json:"startedAt" gorm:"default:null"`
	StoppedAt *time.Time `json:"stoppedAt" gorm:"default:null"`
}

File: web/src/lib/types/vm/vm.ts
import { z } from 'zod/v4';
import { NetworkObjectSchema } from '../network/object';

export interface CreateData {
	name: string;
	node: string;
	id: number;
	description: string;
	storage: {
		type: string;
		guid: string;
		size: number;
		emulation: string;
		iso: string;
	};
	network: {
		switch: number;
		mac: string;
		emulation: string;
	};
	hardware: {
		sockets: number;
		cores: number;
		threads: number;
		memory: number;
		passthroughIds: number[];
		pinnedCPUs: number[];
	};
	advanced: {
		vncPort: number;
		vncPassword: string;
		vncWait: boolean;
		vncResolution: string;
		startAtBoot: boolean;
		bootOrder: number;
		tpmEmulation: boolean;
	};
}

export const VMStorageSchema = z.object({
	id: z.number().int(),
	type: z.string(),
	dataset: z.string(),
	size: z.number().int(),
	emulation: z.string(),
	detached: z.boolean().optional(),
	vmId: z.number().int().optional(),
	bootOrder: z.number().int().optional(),
	name: z.string().optional()
});

export const VMNetworkSchema = z.object({
	id: z.number().int(),
	mac: z.string(),
	macId: z.number().int().optional(),
	macObject: NetworkObjectSchema.optional(),
	switchId: z.number().int(),
	emulation: z.string(),
	vmId: z.number().int().optional()
});

export const VMSchema = z.object({
	id: z.number().int(),
	name: z.string(),
	description: z.string(),
	vmId: z.number().int(),
	cpuSockets: z.number().int(),
	cpuCores: z.number().int(),
	cpuThreads: z.number().int(),
	ram: z.number().int(),
	vncPort: z.number().int(),
	vncPassword: z.string(),
	vncResolution: z.string(),
	vncWait: z.boolean(),
	startAtBoot: z.boolean(),
	startOrder: z.number().int(),
	wol: z.boolean(),

	state: z.enum(['ACTIVE', 'INACTIVE']),

	storages: z.array(VMStorageSchema),
	networks: z.array(VMNetworkSchema),
	pciDevices: z.union([z.array(z.number().int()), z.null()]),
	cpuPinning: z.union([z.array(z.number().int()), z.null()]),

	createdAt: z.string(),
	updatedAt: z.string(),

	startedAt: z.string().nullable(),
	stoppedAt: z.string().nullable()
});

export const VMStatSchema = z.object({
	vmId: z.number().int(),
	cpuUsage: z.number(),
	memoryUsage: z.number(),
	memoryUsed: z.number(),
	createdAt: z.string()
});

export const VMDomainSchema = z.object({
	id: z.number().int(),
	uuid: z.string(),
	name: z.string(),
	status: z.string()
});

export const SimpleVmSchema = z.object({
	id: z.number().int(),
	name: z.string(),
	vmId: z.number().int(),
	state: z.string()
});

export type VM = z.infer<typeof VMSchema>;
export type VMStorage = z.infer<typeof VMStorageSchema>;
export type VMNetwork = z.infer<typeof VMNetworkSchema>;
export type VMDomain = z.infer<typeof VMDomainSchema>;
export type VMStat = z.infer<typeof VMStatSchema>;
export type SimpleVm = z.infer<typeof SimpleVmSchema>;
